scipy/special/_amos.c

```diff
@@ -0,0 +1,6217 @@
+    bk = exp(bk);
+    sum = 0.;
+    km = kk - inu;
+    for (i = 1; i < (km+1); i++)
+    {
+        pt = p2;
+        p2 = p1 + (fkk + fnf)*rz*p2;
+        p1 = pt;
+        ak = 1. - tfnf / (fkk+tfnf);
+        ack = bk*ak;
+        sum += (ack + bk)*p1;
+        bk = ack;
+        fkk -= 1.;
+    }
+    y[n-1] = p2;
+    if (n != 1) {
+        for (i = 2; i < (n+1); i++)
+        {
+            pt = p2;
+            p2 = p1 + (fkk + fnf)*rz*p2;
+            p1 = pt;
+            ak = 1. - tfnf / (fkk+tfnf);
+            ack = bk*ak;
+            sum += (ack + bk)*p1;
+            bk = ack;
+            fkk -= 1.;
+            m = n - i + 1;
+            y[m-1] = p2;
+        }
+    }
+    if (ifnu > 0) {
+        for (i = 1; i < (ifnu+1); i++)
+        {
+            pt = p2;
+            p2 = p1 + (fkk + fnf)*rz*p2;
+            p1 = pt;
+            ak = 1. - tfnf / (fkk+tfnf);
+            ack = bk*ak;
+            sum += (ack + bk)*p1;
+            bk = ack;
+            fkk -= 1.;
+        }
+    }
+    pt = z;
+    if (kode == 2) { pt -= x; }
+    p1 = -fnf * clog(rz) + pt;
+    ap = amos_gamln(1. + fnf);
+    pt = p1 - ap;
+    //
+    // THE DIVISION EXP(PT)/(SUM+P2) IS ALTERED TO AVOID OVERFLOW
+    // IN THE DENOMINATOR BY SQUARING LARGE QUANTITIES
+    //
+    p2 += sum;
+    ap = cabs(p2);
+    p1 = 1. / ap;
+    ck = cexp(pt) * p1;
+    pt = conj(p2)*p1;
+    cnorm = ck * pt;
+    for (int i = 0; i < n; i++) { y[i] *= cnorm; }
+    return nz;
+}
+
+
+int amos_kscl(
+    double complex zr,
+    double fnu,
+    int n,
+    double complex *y,
+    double complex rz,
+    double *ascle,
+    double tol,
+    double elim
+) {
+
+    //***BEGIN PROLOGUE  ZKSCL
+    //***REFER TO  ZBESK
+    //
+    //     SET K FUNCTIONS TO ZERO ON UNDERFLOW, CONTINUE RECURRENCE
+    //     ON SCALED FUNCTIONS UNTIL TWO MEMBERS COME ON SCALE, THEN
+    //     RETURN WITH MIN(NZ+2,N) VALUES SCALED BY 1/TOL.
+    //
+    //***ROUTINES CALLED  ZUCHK,AZABS,AZLOG
+    //***END PROLOGUE  ZKSCL
+
+    double complex cy[2] = { 0. };
+    double as, acs, alas, fn, zri, xx;
+    double complex s1, s2, cs, ck, zd;
+    int nz = 0;
+    int ic = 0;
+    int nn = ( n > 2 ? 2 : n );
+    int kk = 0;
+    int i;
+    double elm = exp(-elim);
+    xx = creal(zr);
+
+    for (i = 1; i < (nn + 1); i++)
+    {
+        s1 = y[i-1];
+        cy[i-1] = s1;
+        as = cabs(s1);
+        acs = -creal(zr) + log(as);
+        nz += 1;
+        y[i-1] = 0.;
+        if (acs < -elim) {
+            continue;
+        }
+        cs = -zr + clog(s1);
+        cs = (exp(creal(cs))/tol)*(cos(cimag(cs)) + sin(cimag(cs)*I));
+        if (!amos_uchk(cs, *ascle, tol)) {
+            y[i-1] = cs;
+            nz -= 1;
+            ic = i;
+        }
+    }
+    if (n == 1) {
+        return nz;
+    }
+    if (ic <= 1) {
+        y[0] = 0.;
+        nz = 2;
+    }
+    if (n == 2) {
+        return nz;
+    }
+    fn = fnu + 1.;
+    ck = fn*rz;
```

Comment:
Maybe there's a missing branch here?
Between L4009~L4010

```c
    if (nz == 0) {
        return nz;
    }
```

https://github.com/scipy/scipy/blob/e5f21298b62fb03732f000645d2caaa67aad8800/scipy/special/amos/zkscl.f#L52-L56

Reply:
I think you are right. Very nice catch. Apologies for missing that. Would you like to send a PR?

Reply:
Yes, I'd be happy to submit a pr, however I'm having some difficulty constructing a test case.
Although the `amos_kscl` function affects almost all the `amos` exported functions, since it is not exported, it can only be tested through the upper level already exported functions.
Currently, the shortest call path seems to be:

```dot
amos_airy 
	-> amos_bknu -> amos_kscl

amos_besh, amos_besk
	-> amos_acon
	-> amos_bknu -> amos_kscl
```

I'm going to open a draft pr and then see what feedback I get from people.

Reply:
Since this is a translation work, it is indeed very challenging to hit certain parts of the codebase unless called specifically internally but I think this case warrants a fix with or without a test case anyways, for obvious reasons. So no problem if you can't find a test case.